### ISSUE TITLE
type 'Null' is not a subtype of type '$Instance' in type cast when using a loop with type declaration

### ISSUE THREAD

**Issue by Noobware1** [ethanblake4/dart_eval#154]
To check at what point the code was throwing the error I added print statements in between the code and found out the error was thrown after "fail 1" also no error was thrown when I replaced the types with var keyword but now even that won't work for me.

```
      String extractedKey = '';
      int currentIndex = 0;
      print('fail 1');
      for (List<int> index in decryptKey) {
        print('fail 2');

        int start = index[0] + currentIndex;
        int end = start + index[1];
        print('fail 3');

        for (int i = start; i < end; i++) {
          print('fail 4');

          extractedKey += sourcesArray[i];
          sourcesArray[i] = '';
          print('fail 5');
        }
        print('fail 6');

        currentIndex += index[1];
        print('fail ');
      }
```

here's the error
```
fail 1
Unhandled exception:
dart_eval runtime exception: type 'Null' is not a subtype of type '$Instance' in type cast
#0      PushObjectProperty.run (package:dart_eval/src/eval/runtime/ops/objects.dart:273:31)
#1      Runtime.bridgeCall (package:dart_eval/src/eval/runtime/runtime.dart:847:12)
#2      Await._suspend (package:dart_eval/src/eval/runtime/ops/bridge.dart:169:13)
at <asynchronous gap>

RUNTIME STATE
=============
Program offset: 277
Stack sample: [L0: Instance of '$ExtractorApi', L1: Instance of '$Completer<dynamic>', L2: $"https://megacloud.tv/embed-2/ajax/e-1/getSources?id=9NU9BtrvgDN6", L3: $"GET", L4: Instance of '$Map<$Value, $Value>', L5: null, L6: Instance of '$Future<dynamic>', L7: Instance of '$OkHttpResponseObject', L8: null, L9: Instance of '$Map<$Value?, $Value?>']
Args sample: []
Call stack: [0, -1]
TRACE:
271: PushConstantInt (0)
272: PushConstant (C34)
273: BoxString (L25)
274: PushArg (L25)
275: InvokeExternal (Ex#0)
276: PushReturnValue ()
277: PushObjectProperty (L17.C35)  <<< EXCEPTION
278: PushReturnValue ()
279: PushNull ()
280: InvokeDynamic (L27.C36)

#0      Runtime.bridgeCall (package:dart_eval/src/eval/runtime/runtime.dart:857:7)
#1      Await._suspend (package:dart_eval/src/eval/runtime/ops/bridge.dart:169:13)
<asynchronous suspension>
```

**Comment by ethanblake4** [comment]
Can you please show the context of the code that comes before this loop? Also what type is `decryptKey`?

(To translate the error, dart_eval is trying to get decryptKey's `iterator` but is finding that `decryptKey` is null)

**Comment by Noobware1** [comment]
![Screenshot_2023-12-11-23-23-00-86_e4424258c8b8649f6e67d283a50a2cbc](https://github.com/ethanblake4/dart_eval/assets/110553648/604ab53c-4d5b-4adc-8c7c-8d23414cbbf3)

Found some old code the current approach is still the same just keyUrl is different and instead of extension methods I am using my own bridged methods

**Comment by Noobware1** [comment]
here's the full code that I am currently using
```
// ignore_for_file: unnecessary_this

import 'dart:convert';

import 'package:meiyou_extenstions/meiyou_extenstions.dart';

// void main(List<String> args) {
//   MegaCloud(ExtractorLink(
//           url: 'https://megacloud.tv/embed-2/e-1/4moscC8xgnRg?k=1', name: ''))
//       .extract()
//       .then(print);
// }

class MegaCloud extends ExtractorApi {
  MegaCloud(ExtractorLink extractorLink) : super(extractorLink);

  String get hostUrl =>
      StringUtils.substringBefore(this.extractorLink.url, '/embed');

  String getUrl() {
    final serverUrl = this.extractorLink.url;
    final embed = RegExp(r'embed-\d').firstMatch(serverUrl)?.group(0);
    final id = StringUtils.substringBefore(
        StringUtils.substringAfterLast(serverUrl, "/"), "?");

    final e = RegExp(r'e-\d').firstMatch(serverUrl)?.group(0) ?? '1';
    return '$hostUrl/$embed/ajax/$e/getSources?id=$id';
  }

  @override
  Future<Video> extract() async {
    final jsonLink = getUrl();

    final response =
        (await AppUtils.httpRequest(url: jsonLink, method: 'GET', headers: {
      'X-Requested-With': 'XMLHttpRequest',
      'Referer': this.extractorLink.url,
      'User-Agent':
          'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:101.0) Gecko/20100101 Firefox/101.0',
    }))
            .json();

    final List<VideoSource> sources;

    if (response['encrypted'] == false) {
      sources = ListUtils.mapList(
          (response['sources'] as List), (e) => toVideoSource(e));
    } else {
      final decryptKey = await _decryptKey();

      List<String> sourcesArray =
          StringUtils.valueToString(response['sources']).split('');

      String extractedKey = '';
      int currentIndex = 0;
      for (var index in decryptKey) {
        int start = index[0] + currentIndex;
        int end = start + index[1];

        for (int i = start; i < end; i++) {
          extractedKey += sourcesArray[i];
          sourcesArray[i] = '';
        }

        currentIndex += index[1];
      }
      final decrypted = CryptoUtils.AES(
          ciphertext: sourcesArray.join(''), key: extractedKey, encrypt: false);

      sources = ListUtils.mapList(
          (json.decode(decrypted) as List), (e) => toVideoSource(e));
    }
    final List<Subtitle>? tracks = response['tracks'] == null
        ? null
        : ListUtils.mapList((response['tracks'] as List), (e) {
            return toSubtitle(e);
          });

    return Video(
      videoSources: sources,
      subtitles: tracks?.where((e) => e.langauge != 'thumbnails').toList(),
    );
  }

  String get keyUrl =>
      'https://raw.githubusercontent.com/Noobware1/zoro-keys/e1/key';

  Future<List<List<int>>> _decryptKey() async {
    return (await AppUtils.httpRequest(url: keyUrl, method: 'GET'))
        .json<List<List<int>>>((json) {
      return ListUtils.mapList(
        json as List,
        (l) => ListUtils.mapList(l, (l) => l as int),
      );
    });
  }

  VideoSource toVideoSource(dynamic e) {
    return VideoSource(
      url: e['file'],
      quality: VideoQuality.hlsMaster,
      format: getVideoFormat(StringUtils.valueToString(e['type'])),
    );
  }

  VideoFormat getVideoFormat(String type) {
    if (type == 'hls') {
      return VideoFormat.hls;
    }
    return VideoFormat.mp4;
  }

  Subtitle toSubtitle(dynamic e) {
    return Subtitle(
      url: e['file'],
      langauge: e['label'] ?? 'thumbnails',
      format: AppUtils.getSubtitleFromatFromUrl(e['file']),
    );
  }

  @override
  String get name => 'MegaCloud';
}
```
and here's the output of the code when running normally in dart
```
Video(mediaType: MediaType.Video, videoSources: [VideoSource(url: https://eb.netmagcdn.com:2228/hls-playback/37e696b4e6141426a49a4bc18e7d2b4fab9ee94d9f7731d36379f0d8013d75e7de744e691f4b088b568735951f50a9d56ce843d105ca328a6bae1c7cc363c692627d2ebe2bf4c837b1b4d756ebf5a60772b6e85c9399b7a7e4df14c35ce423a63febc4833dc7f13aa9fbcea992848660c3fab14160f5548ec29cedd4fbe4f016b0c40b9613047e5442a37107f1544cf1/master.m3u8, format: VideoFormat.hls, quality: VideoQuality(width: 0, height: 0), isBackup: false, title: null)], subtitles: [Subtitle(url: https://ccb.megaresources.co/8f/cc/8fcc85f4449b6422b443ba44bddff8d3/eng-2.vtt, format: SubtitleFormat.vtt, langauge: English, headers: null)], extra: null, headers: null)
```

**Comment by Noobware1** [comment]
btw how did you figure out what the error was saying??? please provide me your wisdom it'll be very helpful for me and kind of you.

**Comment by Noobware1** [comment]
Found the error I was not returning the results from fromJson function
```
 static $Value? $json(Runtime runtime, $Value? target, List<$Value?> args) {
    final fromJson = args[0] as EvalCallable?;
    return (target?.$value as OkHttpResponseObject).json((json) {
      final decoded = json is $Value ? json : runtime.wrapRecursive(json);
      if (fromJson == null) return decoded;
      return fromJson.call(runtime, target, [decoded]);
    });
  }
```

but I am getting a new error now it occurs when I declare type for this  ```for (var index in decryptKey)``` like if I do ```for (List<int> index in decryptKey)```  so I am gonna keep this issue open for now  but other than that everything works fine :)

```
Unhandled exception:
dart_eval runtime exception: type 'int' is not a subtype of type '$Value' in type cast
#0      Unbox.run (package:dart_eval/src/eval/runtime/ops/primitives.dart:335:48)
#1      Runtime.bridgeCall (package:dart_eval/src/eval/runtime/runtime.dart:847:12)
#2      Await._suspend (package:dart_eval/src/eval/runtime/ops/bridge.dart:169:13)
at <asynchronous gap>

RUNTIME STATE
=============
Program offset: 296
Stack sample: [L0: Instance of '$ExtractorApi', L1: Instance of '$Completer<dynamic>', L2: $"https://megacloud.tv/embed-2/ajax/e-1/getSources?id=4qMSiU0DDH77", L3: $"GET", L4: Instance of '$Map<$Value, $Value>', L5: null, L6: Instance of '$Future<dynamic>', L7: Instance of '$OkHttpResponseObject', L8: null, L9: Instance of '$Map<$Value?, $Value?>']
Args sample: []
Call stack: [0, -1]
TRACE:
290: PushConstantInt (1)
291: IndexList (L27[L33])
292: Unbox (L34)
293: NumAdd (L32 + L34)
294: PushNull ()
295: CopyValue (L36 <-- L32)
296: Unbox (L36)  <<< EXCEPTION
297: NumLt (L36 < L35)
298: JumpIfFalse (@318 if L37 == false)
299: Unbox (L23)

#0      Runtime.bridgeCall (package:dart_eval/src/eval/runtime/runtime.dart:857:7)
#1      Await._suspend (package:dart_eval/src/eval/runtime/ops/bridge.dart:169:13)
<asynchronous suspension>
```

**Comment by ethanblake4** [comment]
> btw how did you figure out what the error was saying

Some of it is just from experience, like I know that the bytecode sequence
```
272: PushConstant
273: BoxString (Lx)
274: PushArg (Lx+1)
275: InvokeExternal (Ex#0)
276: PushReturnValue ()
```
always means `print()`. Then I looked in the source code to figure out what happens when you write a for-each loop, and I saw that the first bytecode it generates is to [get the element's iterator] property (https://github.com/ethanblake4/dart_eval/blob/480ce14c62f5f42e4f90c97d1568e10a7cbbc019/lib/src/eval/compiler/statement/for.dart#L29), and since PushObjectProperty was throwing the exception it's clear that's where the issue was. Sorry, it's not a great answer, it's a lot easier for me as the person who wrote the code to do this...

**Comment by Noobware1** [comment]
As someone, who's project is very much dependent on this package it's actually really helpful to understand how it's all really work.

**Comment by ethanblake4** [comment]
Second issue fixed (I think) in v0.7.3
